### ISSUE TITLE
Is the plugin alive? 

### ISSUE THREAD

**Issue by jmgomez** [EpicSkookumScript/SkookumScript-Plugin#47]
The repo does not seem to have too much activity and there is no release for the version 4.25 on the marketplace when 4.26 is about to ship. Is the plugin still under development?

**Comment by error454** [comment]
I still use sk daily across multiple projects but keeping the repo up to date with the latest developments has not been a priority for me. I don't have any visibility into who is using sk, to my knowledge I am the only user and developer.

Releases have become complicated as well since 4.25. There is a critical engine patch that I submitted a pull request for without which often prevents packaged projects that use sk from running at all. So my 4.25 release has been pending epic accepting this pr, which they don't seem to have any intention of doing based on the 3 months of inactivity: https://github.com/EpicGames/UnrealEngine/pull/7077

So things are a bit complicated. Yes the plugin is alive but you need to run a custom engine for it to be stable.

**Comment by AndreaDev237** [comment]
hi @error454  thanks for your reply about the status of sk.
That's unfortunate to hear (about 4.25 problem), 
I was planning to use SK on my new project as main gameplay scripting language in combination with C++ (mainly for optimization and Editor extension) & BP (for artist use) because i'm fascinated by the potential and the ability to be usable for every platform (console included) make it much more attractive.
however some things hold me from do the jump (mainly the abscence of documentation and the risk to become abandoned/completely unmantained)
Know you are still working/using it is a big relief regarding the second problem and i hope the documentation will come back online soon.
However I've two questions,
1: with your patch for 4.25 (referenced here https://github.com/EpicGames/UnrealEngine/pull/7077) the plugin can be used without any other major problem?
Because I still consider to do the jump and try to use it for the entire lenght of the project but I've to be sure to not encounter a dead end in the middle of the development (this can determine the death of the project) 
2: if someone use the plugin and encounter some major problems and is not able to solve them and they will be reported here you will provide some kind of help/support (obviously what you already done it was amazing and you maked this plugin/language still usable and I repeat this is amazing)
Thanks and sorry for the wall of text :D

**Comment by error454** [comment]
> However I've two questions,
> 1: with your patch for 4.25 (referenced here [EpicGames/UnrealEngine#7077](https://github.com/EpicGames/UnrealEngine/pull/7077)) the plugin can be used without any other major problem?

Well it depends on your definition of 'major problem'. What my patch for #7077 fixes is a race condition with engine initialization that can cause sk to get initialized before packages have been loaded. When this happens (and it happens in most packaged projects) sk will be unable to resolve all of the types that are contained in those packages. 

Beyond this, the biggest deficiencies in the language are some quirks with pure blueprint defined types. For instance, the IDE sometimes complains about being unable to see Enums/structs that have been defined via blueprint, but this is just a design-time annoyance. Also blueprint defined booleans are not at feature parity with C++ or sk defined booleans, for example an sk defined boolean can run the coroutine:
`@some_boolean._wait_true`

But if you run the same code on a BP boolean you won't get a syntax error but the code will not execute as expected. Another quirk is passing around class references from BP to sk, things get a little weird, there are bugs filed. So there are a lot of quirks that I haven't had time to diagnose/fix.

> 2: if someone use the plugin and encounter some major problems and is not able to solve them and they will be reported here you will provide some kind of help/support.

I'll always help out if I can as time allows. Supporting the language is a large effort and would easily take up all of my time if I allowed it to. I wish I had more time to dedicate towards improving the product. In my day job, I use sk every day, so I know that the important things are stable, it is a really vital part of my workflow.

Whether you use sk or not is a difficult decision. Personally, I think sk is most useful to me as a fast-prototyping language and as a super debug console. It helps me arrive at the correct solution very quickly and offers many things that blueprints do not, for instance, you cannot even get the unit axis of a transform in blueprints, silly things like that. Many of my sk solutions get moved to C++ afterwards, although some stage-management things are much easier to do in sk so those stay. I also use sk for writing all of my integration tests. It's a really great language and framework with an incredibly difficult and confusing UE4 integration.

So a long answer to your questions. In short, sk has some quirks and if you are coming in new, these quirks might feel like major bugs to you 🤷‍♂️. I had hoped to keep sk going for a long time but progress is slower than I had hoped, although I use sk for professional work, I only fix the showstopper bugs and have a workflow that mitigates the remaining issues. I'd love to get more users and contributors, but also realize there are a lot of onboarding issues in the ecosystem that look unattractive from the outside and those can only be fixed by someone dedicating their unpaid time to work on them.

The spare time I've had in the last few months I was dedicating to improving the IDE autocomplete, as this is an area where I lose a lot of productivity. Again, slower progress than I'd like 😢.

**Comment by jmgomez** [comment]
Thank you for taking the time to reply in depth. It's really appreciated. 

On the pending pull request, can't it be done with a long pull mechanism in place as workaround? I mean, is there any place where you can ask if the maps is already loaded? I told you so because recently I encounter a similar problem with the Houdini Plugin and solved it doing that.

On the use of the lang is self, that's what I looking for a scripting lang on roids not a replacement of cpp or bps.

**Comment by error454** [comment]
> On the pending pull request, can't it be done with a long pull mechanism in place as workaround? I mean, is there any place where you can ask if the maps is already loaded? I told you so because recently I encounter a similar problem with the Houdini Plugin and solved it doing that.

Not easily. The point at which packages are loaded and available is sometime after module initialization and sometime before the world is created and `BeginPlay` is called on actors. If `BeginPlay` is called before we've done our init then we crash. Tick events don't start going until after the world is created, so we can't simply check things every tick. 

If we spin up a new thread and check there periodically then we have to worry about data synchronization and thread safety and I don't see any existing thread safe variables or mutexes designed to do this. The patch I submitted (a measly 4 lines of code) is the best way of doing this IMO, it is straight-forward and does exactly what is required without any extra mess.

**Comment by AndreaDev237** [comment]
> > However I've two questions,
> > 1: with your patch for 4.25 (referenced here [EpicGames/UnrealEngine#7077](https://github.com/EpicGames/UnrealEngine/pull/7077)) the plugin can be used without any other major problem?
> 
> Well it depends on your definition of 'major problem'. What my patch for #7077 fixes is a race condition with engine initialization that can cause sk to get initialized before packages have been loaded. When this happens (and it happens in most packaged projects) sk will be unable to resolve all of the types that are contained in those packages.
> 
> Beyond this, the biggest deficiencies in the language are some quirks with pure blueprint defined types. For instance, the IDE sometimes complains about being unable to see Enums/structs that have been defined via blueprint, but this is just a design-time annoyance. Also blueprint defined booleans are not at feature parity with C++ or sk defined booleans, for example an sk defined boolean can run the coroutine:
> `@some_boolean._wait_true`
> 
> But if you run the same code on a BP boolean you won't get a syntax error but the code will not execute as expected. Another quirk is passing around class references from BP to sk, things get a little weird, there are bugs filed. So there are a lot of quirks that I haven't had time to diagnose/fix.
> 
> > 2: if someone use the plugin and encounter some major problems and is not able to solve them and they will be reported here you will provide some kind of help/support.
> 
> I'll always help out if I can as time allows. Supporting the language is a large effort and would easily take up all of my time if I allowed it to. I wish I had more time to dedicate towards improving the product. In my day job, I use sk every day, so I know that the important things are stable, it is a really vital part of my workflow.
> 
> Whether you use sk or not is a difficult decision. Personally, I think sk is most useful to me as a fast-prototyping language and as a super debug console. It helps me arrive at the correct solution very quickly and offers many things that blueprints do not, for instance, you cannot even get the unit axis of a transform in blueprints, silly things like that. Many of my sk solutions get moved to C++ afterwards, although some stage-management things are much easier to do in sk so those stay. I also use sk for writing all of my integration tests. It's a really great language and framework with an incredibly difficult and confusing UE4 integration.
> 
> So a long answer to your questions. In short, sk has some quirks and if you are coming in new, these quirks might feel like major bugs to you 🤷‍♂️. I had hoped to keep sk going for a long time but progress is slower than I had hoped, although I use sk for professional work, I only fix the showstopper bugs and have a workflow that mitigates the remaining issues. I'd love to get more users and contributors, but also realize there are a lot of onboarding issues in the ecosystem that look unattractive from the outside and those can only be fixed by someone dedicating their unpaid time to work on them.
> 
> The spare time I've had in the last few months I was dedicating to improving the IDE autocomplete, as this is an area where I lose a lot of productivity. Again, slower progress than I'd like 😢.

thanks for the amazing reply,
This will show again your commitment to this and give me a huge motivational boost,
Is true there are some quirks but i think the benefits overcome the issues so i'm still motivated to try it,
Thanks also for the explanation about your usage of sk inside a project this will be useful also for me,
I think i will download the latest version of the engine and i will apply your patch on it (i think even in the next version of the engine this will be easy mergable) 
I don't know why epic is so slow with this easy fix.
Regarding the help about the sk mantainment and improvement if i will implement it succesfully inside a project and it will become my standard working bench i will try to do my best to help when i can! :D
Thank you again for everything

**Comment by Noolarch** [comment]
Hopefully you are following the [SkookumScript forum](https://skookum.chat/t/documentation-offline/1992/15?u=noolarch) though in the vein of the plugin being alive - the [SkookumScript web site](https://skookumscript.com/) is finally back!

The issue holding up the Sk plugin in the more recent UE versions is being addressed at Epic. We will note whenever any related news is available.
